lib/Catalyst/Action.pm

```diff
@@ -39,38 +39,73 @@ has private_path => (
   default => sub { '/'.shift->reverse },
 );
 
-has args_constraints => (
-  is=>'ro',
-  traits=>['Array'],
-  isa=>'ArrayRef',
-  required=>1,
-  lazy=>1,
-  builder=>'_build_args_constraints',
-  handles => {
-    has_args_constraints => 'count',
-    number_of_args => 'count',
-    all_args_constraints => 'elements',
-  });
-
-  sub _build_args_constraints {
+has number_of_args => (
+    is => 'rw',
+    lazy => 1,
+    builder=>'_build_number_of_args',
+);
+
+has ['preargs', 'postargs'] => (is => 'rw', default => sub { [] } );
+
+
+  sub _build_number_of_args {
     my $self = shift;
+
     my @arg_protos = @{$self->attributes->{Args}||[]};
 
-    return [] unless scalar(@arg_protos);
-    # If there is only one arg and it looks like a number
-    # we assume its 'classic' and the number is the number of
-    # constraints.
+    my @extracted;
+    for my $arg (@arg_protos) {
+        if (defined $arg) {
+            push( @extracted, map { $_ eq '*' ? undef : $_ } split(/\s*,\s*/, $arg) );
+        }
+        else {
+            push( @extracted, $arg );
+        }
+    }
+    @arg_protos = @extracted;
+
+    return undef unless @arg_protos;
+
     my @args = ();
-    if(
-      scalar(@arg_protos) == 1 &&
-      looks_like_number($arg_protos[0])
-    ) {
-      return [];
-    } else {
-      @args = map { Moose::Util::TypeConstraints::find_or_parse_type_constraint($_) || die "$_ is not a constraint!" } @arg_protos;
+    my ($preany, $any, $postany) = (0,0,0);
+    for (0 .. $#arg_protos) {
+        if ( defined( $arg_protos[$_] ) && looks_like_number( $arg_protos[$_] ) ) {
+            push( @args, 0 + $arg_protos[$_] );
+            $preany += $arg_protos[$_];
+        }
+        elsif ( defined $arg_protos[$_] ) {
+            my $constraint = Moose::Util::TypeConstraints::find_or_parse_type_constraint($arg_protos[$_]) or die "$arg_protos[$_] is not a constraint!";
+            push(@args, $constraint);
+            $preany += 1;
+        }
+        else {
+            $any = 1;
+            last;
+        }
+    }
+
+    my @postargs = ();
+    if ($any) {
+        for (reverse(0 .. $#arg_protos)) {
+            if ( defined( $arg_protos[$_] ) && looks_like_number( $arg_protos[$_] ) ) {
+                push( @postargs, 0 + $arg_protos[$_] );
+                $postany += $arg_protos[$_];
+            }
+            elsif ( defined $arg_protos[$_] ) {
+                my $constraint = Moose::Util::TypeConstraints::find_or_parse_type_constraint($arg_protos[$_]) or die "$arg_protos[$_] is not a constraint!";
+                push(@postargs, $constraint);
+                $postany += 1;
+            }
+            else {
+                last;
+            }
+        }
     }
+    #warn "$preany, $any, $postany";
+    $self->preargs([@args]);
+    $self->postargs([@postargs]);
```

Comment:
I think we need to find a way to not need mutable attributes.  The Action is a singleton and associated I think at the application not context level ( i need to double check ) so we could get leakage here across requests.  The rest of it is looking good!